cni.go

```diff
@@ -0,0 +1,138 @@
+//
+// Copyright (c) 2016 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package virtcontainers
+
+import (
+	"github.com/containernetworking/cni/pkg/ns"
+	cniPlugin "github.com/containers/virtcontainers/network/cni"
+	"github.com/golang/glog"
+)
+
+// cni is a network implementation for the CNI plugin.
+type cni struct{}
+
+func (n *cni) addVirtInterfaces(config NetworkConfig, netPairs []NetworkInterfacePair) error {
+	netPlugin, err := cniPlugin.NewNetworkPlugin()
+	if err != nil {
+		return err
+	}
+
+	for _, pair := range netPairs {
+		res, err := netPlugin.AddNetwork(pair.ID, config.NetNSPath, pair.VirtIface.Name)
+		if err != nil {
+			return err
+		}
+
+		glog.Infof("AddNetwork results %v\n", res)
+	}
+
+	return nil
+}
+
+func (n *cni) deleteVirtInterfaces(config NetworkConfig, netPairs []NetworkInterfacePair) error {
+	netPlugin, err := cniPlugin.NewNetworkPlugin()
+	if err != nil {
+		return err
+	}
+
+	for _, pair := range netPairs {
+		err := netPlugin.RemoveNetwork(pair.ID, config.NetNSPath, pair.VirtIface.Name)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// add creates a new network namespace and its virtual network interfaces,
+// and it creates and bridges TAP interfaces for the CNI network.
+func (n *cni) add(config *NetworkConfig) ([]NetworkInterfacePair, error) {
```

Comment:
I'd like this one to return more details, i.e. a complete networking namespace description.

```Go
type Endpoint struct {
        netPair    NetworkInterfacePair
        properties types.Result
}

type NetworkNamespace struct {
       netNsPath string
       endpoints  []Endpoint
}
```

I know this is bit CNI centric, but `types.Result` is actually network generic, and we can build it from a networking namespace scan with CNM.

So `add()` would return a `NetworkNamespace` and we would add that to the created pod.
The storing the pod would be split into storing the config and storing the whole `NetworkNamespace` into 2 different files.